File: 백준/Gold/6156. Cow Contest/Cow Contest.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.stream.IntStream;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] nm = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int n = nm[0];
        int m = nm[1];

        boolean[][] graph = new boolean[n][n];
        while (m-- > 0) {
            int[] ab = Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();

            int a = ab[0];
            int b = ab[1];

            graph[a-1][b-1] = true;
        }

        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (graph[i][k] && graph[k][j])
                        graph[i][j] = true;
                }
            }
        }

        int count = 0;
        for (int i = 0; i < n; i++) {
            int finalI = i;
            int wins = IntStream.range(0, n)
                    .map(j -> graph[finalI][j] ? 1 : 0)
                    .sum();

            int finalI1 = i;
            int loses = IntStream.range(0, n)
                    .map(j -> graph[j][finalI1] ? 1 : 0)
                    .sum();

            count += (wins + loses == n - 1) ? 1 : 0;
        }

        System.out.println(count);
    }
}

File: 백준/Gold/6156. Cow Contest/Cow Contest.py
# https://www.acmicpc.net/problem/6156
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
graph = [[0] * (n+1) for _ in range(n+1)]
for _ in range(m):
    a, b = map(int, input().split())
    graph[a][b] = 1

for k in range(1, n+1):
    for i in range(1, n+1):
        for j in range(1, n+1):
            if graph[i][k] and graph[k][j]:
                graph[i][j] = 1

ans = 0
for i in range(1, n+1):
    wins = sum(graph[i])
    loses = sum(graph[j][i] for j in range(1, n+1))

    # 나에게 이기는 수와 진 수의 합이 n-1인 경우 내 순위는 정해진 것임
    if wins + loses == n - 1:
        ans += 1

print(ans)